### DISCUSSION TITLE
Need sample workflow definition for HTTP POST Request (which has request payload as json data)

### DISCUSSION THREAD

**Discussion by karthanvijay** [serverlessworkflow/synapse#175] [Q&A] [answered]
Need sample workflow definition for HTTP POST Request (which has request payload as json data) 
Below is my sample workflow definition
 ```json
{
  "expressionLang": "jq",
  "functions": [
    {
      "name": "FeatureapiFeaturePOST",
      "operation": "http://*.*.*.*:80/swagger/v1/swagger.json#FeatureapiFeatureSave",
      "type": "rest",
      "authRef": "bearer_token"
    }
  ],
  "auth": [
    {
      "scheme": "bearer",
      "properties": {
        "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJFbXBsb3llZUlEIjoiMTAwMDI1OTYxNyIsIlVzZXJOYW1lIjoiS2FydGhhbiBWaWpheSIsIkVtYWlsIjoiS2FydGhhbi5WaWpheUB3ZGMuY29tIiwiZXhwIjoxNjYwNjQ4MDI2LCJpc3MiOiJDb25maWcgRWNvc3lzdGVtIiwiYXVkIjoiQ29uZmlnIEVjb3N5c3RlbSJ9.XVv8WMwDblsVN8cmlSOPiMC5dmTZX00EP28nvNFH49s"
      },
      "name": "bearer_token"
    }
  ],
  "id": "AuthTokenPostSampleRequest",
  "name": "Test",
  "version": "1.0.5",
  "specVersion": "0.8",
  "keepActive": false,
  "states": [
    {
      "end": true,
      "actionMode": "sequential",
      "type": "operation",
      "actions": [
        {
          "functionRef": {
            "refName": "FeatureapiFeaturePOST",
            "arguments": {
              "inputData": "${. }"
            }
          },
          "name": "OpenAPI"
        }
      ],
      "name": "CofigSettings",
      "usedForCompensation": false
    }
  ],
  "autoRetries": false
}
```

**Comment by JBBianchi** [comment, marked as answer]
If you want to use the workflow's input as argument for a POST, you can use `body`. Here is a petstore sample:

```json
{
   "expressionLang":"jq",
   "functions":[
      {
         "name":"AddPetFn",
         "operation":"https://petstore.swagger.io/v2/swagger.json/#addPet",
         "type":"rest"
      }
   ],
   "id":"addpet",
   "name":"Add Pet",
   "version":"1.0.0",
   "specVersion":"0.8",
   "keepActive":false,
   "states":[
      {
         "name":"AddPet",
         "end":true,
         "type":"operation",
         "actions":[
            {
               "name":"addpetaction",
               "functionRef":{
                  "refName":"AddPetFn",
                  "arguments":{
                     "body":"${. }"
                  }
               }
            }
         ]
      }
   ]
}
```

**Reply by karthanvijay** [reply]
Hi @JBBianchi,

Thanks for the reply, I have made above changes, I am seeing below Error:
 ```json
{
  "code": "",
  "message": "An error occured while evaluting the specified expression: bash: line 1: jq.exe: command not found\ncat: write error: Broken pipe\n"
}  
```

**Reply by cdavernas** [reply]
How did you install Synapse? If you did it manually, in self hosted mode, then you need to install `jq`, too. You need to add it to your path environement variables too.

You can find jq [here](https://stedolan.github.io/jq/download/)

**Reply by karthanvijay** [reply]
Hi @cdavernas , 

How did you install Synapse? --> I am running Synapse instance in docker, Do I need to run  jq as well in docker?

I have couple of questions as well:

1. How to pass bearer_token as part of request instead of defining in workflow json?
2. In HttpPost request call, request body json is getting converted to camel case is this default behaviour?
3. In HttpPost request call, If request body is huge json it throws error saying too many parameters, is there any limitation in request body length(size/parameters)?

4.  Condition expression is failing to evaluate with below error Workflow json shared below {
  "code": "NullReference",
  "message": "Failed to find a condition with the specified name ''"
}
Workflow json:
```json
{
    "expressionLang": "jq",
     "start": "ContributorCommit",
    "functions": [{
            "name": "ApprovalFeaturePOST",
            "operation": "http://*.*.*.*:80/swagger/v1/swagger.json#ApprovalapiStagingSaveMultiple",
            "type": "rest",
            "authRef": "bearer_token"
        }
    ],
    "auth": [{
            "scheme": "bearer",
            "properties": {
                "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJFbXBsb3llZUlEIjoiMTAwMDI1OTYxNyIsIlVzZXJOYW1lIjoiS2FydGhhbiBWaWpheSIsIkVtYWlsIjoiS2FydGhhbi5WaWpheUB3ZGMuY29tIiwiZXhwIjoxNjYxMTc2MDE3LCJpc3MiOiJDb25maWcgRWNvc3lzdGVtIiwiYXVkIjoiQ29uZmlnIEVjb3N5c3RlbSJ9.Llk3DcEcT-dkqcFYE0ZGLFs5K2seUSEKZoZTqRYj7QA"
            },
            "name": "bearer_token"
        }
    ],
    "id": "Feature_AutoApproval",
    "name": "Test",
    "version": "1.0.7",
    "specVersion": "0.8",
    "keepActive": false,
    "states": [{
            "dataConditions": [{
                    "transition": "AutoApproval",
                    "condition": "${ .role == \"Admin\" }"
                }
            ],
            "type": "switch",
            "defaultCondition": {
                "transition": "Completed"
            },
            "name": "DetermineRole",
            "usedForCompensation": false
        }, {
            "end": true,
            "data": {
                "result": "Task Completed!"
            },
            "type": "inject",
            "name": "Completed",
            "usedForCompensation": false
        }, {
            "transition": "Completed",
            "actionMode": "sequential",
            "type": "operation",
            "actions": [{
                    "subFlowRef": {
                        "invoke": "async",
                        "workflowId": "Feature_Approval",
                        "version": "1.0.1"
                    },
                    "name": "AutoAprrovalFlow"
                }
            ],
            "name": "AutoApproval",
            "usedForCompensation": false
        }, {
            "stateDataFilter": {
                "output": "${. }"
            },
            "transition": "DetermineRole",
            "actionMode": "sequential",
            "type": "operation",
            "actions": [{
                    "subFlowRef": {
                        "invoke": "async",
                        "workflowId": "Feature_ContributorCommit",
                        "version": "1.0.7"
                    },
                    "name": "ContributorCommitFlow",
                    "actionDataFilter": {
                        "results": "${. }",
                        "useResults": true
                    }
                }
            ],
            "name": "ContributorCommit",
            "usedForCompensation": false
        }
    ],
    "autoRetries": false
}
```

**Comment by cdavernas** [comment]
Hi @karthanvijay!

> I am running Synapse instance in docker, Do I need to run jq as well in docker?

No, you don't. Your issue might then be related to #179, which I coincidentally ran into last week. It has been fixed in [release 0.1.15](https://github.com/serverlessworkflow/synapse/releases/tag/0.1.15).

> How to pass bearer_token as part of request instead of defining in workflow json?

Defining it in the workflow allows for transparent use of your token, which will be added to both the `OpenAPI` file request, and the described operation's. Just as you did in the above example definition.

At the moment, there's no "out-of-the-box" way to add parameters that are not explicitely defined by the `OpenAPI` document. So, if you want to add it explicitely via a `functionRef`'s `arguments` property, you need to define it as a `in-header` parameter in your `OpenAPI` document.

If you want to pass your token to a `subflowRefs`, however, you'd need to explicitely add it to the state data, either with an inject state, or using a `jq` expression in a state/action data filter.

> In HttpPost request call, request body json is getting converted to camel case is this default behaviour?

No, that is a bug I created #192 for. It will be fixed by tomorrow.

> Condition expression is failing to evaluate with below error Workflow json shared below

This is normal, though there's no way you could know about it, as the SW spec doesn't enforce some of Synapse's rules. 

In this case, it is because Synapse expects all conditions to be named, which is anyways a good practice as the workflow's representation becomes more ubiquitous. By the way, in SW `0.9`, the default condition is nameable too, which is not the case in `0.8`

I hope this answer your questions ;)

**Reply by karthanvijay** [reply]
many thanks @cdavernas , your turn around time to answer the queries are much appreciated

In HttpPost request call, If request body is huge json it throws error saying too many parameters, is there any limitation in request body length(size/parameters)?

**Reply by cdavernas** [reply]
@karthanvijay with pleasure! Don't hesitate if you have other questions 😉

**Reply by karthanvijay** [reply]
sure thanks, I think you missed answering this query "In HttpPost request call, If request body is huge json it throws error saying too many parameters, is there any limitation in request body length(size/parameters)?"

**Reply by cdavernas** [reply]
@karthanvijay Indeed, sorry! There should be no limitations, so this might be related to a shell-related issue I've fixed with `0.1.15` too. 

Try latest version and, if it persists, you can open an issue, providing sample and stacktrace and I'll look into it

**Reply by karthanvijay** [reply]
sure @cdavernas thanks

**Comment by cdavernas** [comment]
@karthanvijay Btw, don't hesitate to drop a star to show support to the project 😉
